File: pythonPackages/ufpy/NotificationMessage.py
##
# This software was developed and / or modified by Raytheon Company,
# pursuant to Contract DG133W-05-CQ-1067 with the US Government.
# 
# U.S. EXPORT CONTROLLED TECHNICAL DATA
# This software product contains export-restricted data whose
# export/transfer/disclosure is restricted by U.S. law. Dissemination
# to non-U.S. persons whether in the United States or abroad requires
# an export license or other authorization.
# 
# Contractor Name:        Raytheon Company
# Contractor Address:     6825 Pine Street, Suite 340
#                         Mail Stop B8
#                         Omaha, NE 68106
#                         402.291.0100
# 
# See the AWIPS II Master Rights File ("Master Rights File.pdf") for
# further licensing information.
##



import socket
import sys

from . import ThriftClient
from dynamicserialize.dstypes.com.raytheon.uf.common.alertviz import AlertVizRequest

#
# Provides a capability of constructing notification messages and sending 
# them to a STOMP data source.
#  
#    
#     SOFTWARE HISTORY
#    
#    Date            Ticket#       Engineer       Description
#    ------------    ----------    -----------    --------------------------
#    09/30/08                      chammack       Initial Creation.
#    11/03/10        5849          cjeanbap       Moved to ufpy package from
#                                                     com.raytheon.uf.tools.cli
#    01/07/11        5645          cjeanbap       Added audio file to Status Message.
#    05/27/11        3050          cjeanbap       Added if-statement to check Priority
#                                                 value
#    07/27/15        4654          skorolev       Added filters
#    11/11/15        5120          rferrel        Cannot serialize empty filters.
#    03/05/18        6899          dgilling       Update to latest version of stomp.py API.
#    09/14/18        7464          dgilling       Only serialize audioFile if provided.
#    Apr 16, 2020    8144          tgurney        Change AlertViz stomp port to
#                                                 calculated value based on uid
#    May 15, 2020    8144          tgurney        Remove local-delivery logic
#                                                 (no longer works as of 19.3.4)

class NotificationMessage:

   priorityMap = { 
             0: 'CRITICAL',
             1: 'SIGNIFICANT',
             2: 'PROBLEM',
             3: 'EVENTA',
             4: 'EVENTB',
             5: 'VERBOSE'}

   def __init__(self, host='localhost', port=9581, message='', priority='PROBLEM', category="LOCAL", source="ANNOUNCER", audioFile=None, filters=None):
      self.host = host
      self.port = port
      self.message = message
      self.audioFile = audioFile
      self.source = source
      self.category = category 
      self.filters = filters

      priorityInt = None

      try:
         priorityInt = int(priority)
      except:
         pass

      if priorityInt is None:
          #UFStatus.java contains mapping of Priority to Logging level mapping
          if priority == 'CRITICAL' or priority == 'FATAL':
            priorityInt = int(0)
          elif priority == 'SIGNIFICANT' or priority == 'ERROR':
              priorityInt = int(1)
          elif priority == 'PROBLEM' or priority == 'WARN':
              priorityInt = int(2)
          elif priority == 'EVENTA' or priority == 'INFO':
              priorityInt = int(3)
          elif priority == 'EVENTB':
              priorityInt = int(4)
          elif priority == 'VERBOSE' or priority == 'DEBUG':
              priorityInt = int(5)

      if (priorityInt < 0 or priorityInt > 5):
          print("Error occurred, supplied an invalid Priority value:", str(priorityInt))
          print("Priority values are 0, 1, 2, 3, 4 and 5.")
          sys.exit(1)

      if priorityInt is not None:
          self.priority = self.priorityMap[priorityInt]
      else:
          self.priority = priority

   def send(self):
        alertVizRequest = createRequest(self.message, self.priority, self.source, self.category, self.audioFile, self.filters)
        thriftClient = ThriftClient.ThriftClient(self.host, self.port, "/services")

        serverResponse = None
        try:
            serverResponse = thriftClient.sendRequest(alertVizRequest)
        except Exception as ex:
            print("Caught exception submitting AlertVizRequest:", str(ex))    

        if (serverResponse != "None"):
            print("Error occurred submitting Notification Message to AlertViz receiver:", serverResponse)
            sys.exit(1)
        else:
            print("Response:", str(serverResponse))        

def createRequest(message, priority, source, category, audioFile, filters):    
    obj = AlertVizRequest()

    obj.setMachine(socket.gethostname())    
    obj.setPriority(priority)
    obj.setCategory(category)
    obj.setSourceKey(source)    
    obj.setMessage(message)
    if (audioFile is not None):
        obj.setAudioFile(audioFile)
    else:
        obj.setAudioFile('\0')
    obj.setFilters(filters)
    return obj

if __name__ == '__main__':
    main()

File: cave/com.raytheon.uf.viz.core/src/com/raytheon/uf/viz/core/notification/NotificationMessage.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.uf.viz.core.notification;

import java.util.ArrayList;
import java.util.Enumeration;
import java.util.List;

import javax.jms.BytesMessage;
import javax.jms.JMSException;
import javax.jms.Message;
import javax.jms.TextMessage;

import com.raytheon.uf.common.serialization.DynamicSerializationManager;
import com.raytheon.uf.common.serialization.DynamicSerializationManager.SerializationType;
import com.raytheon.uf.common.serialization.SerializationUtil;
import com.raytheon.uf.viz.core.exception.VizException;
import com.raytheon.uf.common.util.DataUnzipper;

/**
 * Encapsulation object for notification messages
 * 
 * <pre>
 * SOFTWARE HISTORY
 * Date         Ticket#    Engineer    Description
 * ------------ ---------- ----------- --------------------------
 * Sep 2, 2008  1448       chammack     Initial creation
 * Oct 4, 2010  7193       cjeanbap     Added a new method, isNotExpired().
 * Feb 1, 2011  7193       cjeanbap     Added a new method, getPublishedTime().
 * Aug 6, 2013  2228       njensen      Use deserialize(byte[])
 * Aug 16, 2013 2169       bkowal       Unzip any gzipped information
 * </pre>
 * 
 * @author chammack
 * @version 1.0
 */

public class NotificationMessage {

    /** The "raw" jms message */
    private Message jmsMessage;

    /** An unmarshalled object cached for future use */
    private Object unmarshalledObject;

    /**
     * Construct a notification message from a JMS message
     * 
     * @param message
     *            the jms message
     */
    public NotificationMessage(Message message) {
        this.jmsMessage = message;
        if (!(message instanceof TextMessage)
                && !(message instanceof BytesMessage))
            throw new IllegalArgumentException("Unsupported message type: "
                    + message);
    }

    /**
     * Return the unmarshalled object
     * 
     * NOTE: This returns the common message object. Do NOT modify this object.
     * If you need to modify it, make a copy first.
     * 
     * @return the unmarshalled object
     */
    public Object getMessagePayload() throws NotificationException {

        if (this.unmarshalledObject == null) {
            synchronized (jmsMessage) {
                try {
                    // Support messages serialized in binary or XML format
                    if (this.jmsMessage instanceof BytesMessage) {
                        BytesMessage bytesMessage = (BytesMessage) this.jmsMessage;
                        bytesMessage.reset();
                        long length = bytesMessage.getBodyLength();
                        byte[] data = new byte[(int) length];

                        int readLength = bytesMessage.readBytes(data);
                        // QC the data length
                        if (length != readLength)
                            throw new NotificationException(
                                    "Message payload terminated early.  Expected: "
                                            + length + ".  Got: " + readLength);
                        if (DataUnzipper.isGzipped(data)) {
                            data = new DataUnzipper().gunzip(data);
                        }
                        this.unmarshalledObject = DynamicSerializationManager
                                .getManager(SerializationType.Thrift)
                                .deserialize(data);
                    } else if (this.jmsMessage instanceof TextMessage) {
                        TextMessage textMessage = (TextMessage) this.jmsMessage;
                        this.unmarshalledObject = SerializationUtil
                                .unmarshalFromXml(textMessage.getText());
                    }
                } catch (Exception e) {
                    throw new NotificationException(
                            "Error deserializing message payload", e);
                }
            }
        }
        return this.unmarshalledObject;
    }

    /**
     * Return a list of properties
     * 
     * @return a list of properties
     * @throws VizException
     */
    public List<String> getProperties() throws NotificationException {
        try {
            Enumeration<?> enumer = this.jmsMessage.getPropertyNames();

            List<String> retVal = new ArrayList<String>();
            while (enumer.hasMoreElements()) {
                retVal.add((String) enumer.nextElement());
            }

            return retVal;
        } catch (JMSException e) {
            throw new NotificationException("Error reading property names", e);
        }
    }

    /**
     * Get a property by name
     * 
     * @param propertyName
     *            the property name
     * @return the property value
     * @throws VizException
     */
    public Object getProperty(String propertyName) throws NotificationException {
        try {
            return this.jmsMessage.getObjectProperty(propertyName);
        } catch (JMSException e) {
            throw new NotificationException("Error reading property", e);
        }
    }

    /**
     * Return the source information (where the message came from)
     * 
     * @return the source information
     * @throws VizException
     */
    public String getSource() throws NotificationException {
        try {
            return this.jmsMessage.getJMSDestination().toString();
        } catch (JMSException e) {
            throw new NotificationException(
                    "Error retrieving source information", e);
        }
    }

    /**
     * Returns true, if the message has not expired set by the time-to-live
     * value on the Message object; otherwise false.
     * 
     * @return true, if the message has not expired.
     * @throws NotificationException
     */
    public boolean isNotExpired() throws NotificationException {
        try {
            long currentTime = System.currentTimeMillis();
            return (currentTime < this.jmsMessage.getJMSExpiration() ? true
                    : false);
        } catch (JMSException e) {
            throw new NotificationException(
                    "Error retrieving source information", e);
        }
    }

    /**
     * Returns the time as a long when the message was handled off to the
     * provider to be sent.
     * 
     * @return long, the time as in milliseconds.
     * @throws NotificationException
     */
    public long getPublishedTime() throws NotificationException {
        try {
            return this.jmsMessage.getJMSTimestamp();
        } catch (JMSException e) {
            throw new NotificationException(
                    "Error retrieving source information", e);
        }
    }
}
